job posting https://www.indeed.com/viewjob?from=app-tracker-saved-appcard&hl=en&jk=e7942242cccbb791&tk=1go5dqjvch7j1800

## Atmosfy Analysis
### By: Tyler Thomas

## Atmosfy, Inc is looking for a Data Analyst for their remote team. The role requires conducting analysis to support the Product Team and developing insights for the product roadmap

## Atmosfy is a platform for live video of dining and nightlife that helps people discover the world's best experiences and supports local businesses globally by sharing unique experiences through video

In [ ]:
pip install mysql-connector-python

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.7/43.7 kB 2.4 MB/s eta 0:00:00


In [ ]:
pip install --upgrade 'sqlalchemy<2.0' mysql-connector-python

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
import pandas as pd
import mysql.connector
from sqlalchemy import create_engine

In [ ]:
#connect to aws
engine = create_engine('mysql://USERNAME:PASSWORD@HOST/DATABASE')


In [ ]:
#make sure connection is valid
engine

Engine(mysql+mysqlconnector://admin:***@isba-dev-01.cttrjme1cih6.us-east-1.rds.amazonaws.com/SQL Project)

In [ ]:
#load 3 tables from aws into notebook
df1 = pd.read_sql('SELECT * FROM Twitter_Reviews', con=engine)
#any.do
df2 = pd.read_sql('SELECT * FROM playstore_reviews', con=engine)
df3 = pd.read_sql('SELECT * FROM tiktok_google_play_reviews', con=engine)

In [ ]:
df1

Review_Id       User_Name  \
0     d80a474e-421c-408e-aa49-b06185a3830a    Sean Rollins   
1     adce2dfc-11d4-43e0-b4b3-891d942a3457     Andrew Pike   
2     205b4fe2-50f6-4cbd-b63a-d6a3722dfa24    Dean Hampton   
3     2b18dec4-3094-4539-8f6e-dc51fe62fb21            Dave   
4     426d8c6f-df41-459d-936c-eea1154af0b8    Royce Miller   
...                                    ...             ...   
1193  6385bae0-3945-48ae-b2fe-f7db9f5233a4      حسین طاهری   
1194  c8147b0c-cded-4c16-b7b4-0b9b79845677    Ye Mann Aung   
1195  cad48954-2dc1-49d0-ad38-aa2dbc367b29   Kazibwe Tonny   
1196  50130c50-99aa-48e6-ac49-7cb9b4c704e7  Mandeep Jareda   
1197  1e1abc05-53c6-4f21-9fbc-28dd5243b50f    chandra bhan   

                                             User_Image  \
0     https://play-lh.googleusercontent.com/a-/ACNPE...   
1     https://play-lh.googleusercontent.com/a-/ACNPE...   
2     https://play-lh.googleusercontent.com/a/ALm5wu...   
3     https://play-lh.googleusercontent.com/a/ALm5wu...   
4     https://play-lh.googleusercontent.com/a-/ACNPE...   
...                                                 ...   
1193  https://play-lh.googleusercontent.com/a-/ACNPE...   
1194  https://play-lh.googleusercontent.com/a/ALm5wu...   
1195  https://play-lh.googleusercontent.com/a/ALm5wu...   
1196  https://play-lh.googleusercontent.com/a-/ACNPE...   
1197  https://play-lh.googleusercontent.com/a-/ACNPE...   

                                                Comment  Thumbs_Up_Count  \
0     Twitter's obsession with turning itself into t...             7121   
1     Absolute rubbish. It still keeps refreshing af...             4833   
2     The most recent update has ruined this app. Th...             2409   
3     No closed caption (cc) button is present to di...             4033   
4     This app is frustrating as heck. When you're s...             1671   
...                                                 ...              ...   
1193                                              graet                1   
1194                                             I like                1   
1195                                          Great app                1   
1196                            5 star for Elon Musk ✌️                2   
1197                                           Nice app                0   

               Version Review_Created_At     Sort_Order  
0     9.65.0-release.0  28-10-2022 01:48  most_relevant  
1     9.56.2-release.0  26-08-2022 08:26  most_relevant  
2     9.56.2-release.0  26-08-2022 18:31  most_relevant  
3     9.56.2-release.0  26-08-2022 16:05  most_relevant  
4     9.57.0-release.1  02-09-2022 07:37  most_relevant  
...                ...               ...            ...  
1193  9.65.1-release.0  05-11-2022 11:56         newest  
1194  9.65.1-release.0  05-11-2022 11:54         newest  
1195                    05-11-2022 11:53         newest  
1196  9.65.1-release.0  05-11-2022 11:53         newest  
1197  9.61.0-release.0  05-11-2022 11:47         newest  

[1198 rows x 8 columns]

In [ ]:
#removing unwanted columns
df1 = df1.drop(columns=['Sort_Order'])

In [ ]:
df1 = df1.drop(columns=['User_Image'])

In [ ]:
df1

Review_Id       User_Name  \
0     d80a474e-421c-408e-aa49-b06185a3830a    Sean Rollins   
1     adce2dfc-11d4-43e0-b4b3-891d942a3457     Andrew Pike   
2     205b4fe2-50f6-4cbd-b63a-d6a3722dfa24    Dean Hampton   
3     2b18dec4-3094-4539-8f6e-dc51fe62fb21            Dave   
4     426d8c6f-df41-459d-936c-eea1154af0b8    Royce Miller   
...                                    ...             ...   
1193  6385bae0-3945-48ae-b2fe-f7db9f5233a4      حسین طاهری   
1194  c8147b0c-cded-4c16-b7b4-0b9b79845677    Ye Mann Aung   
1195  cad48954-2dc1-49d0-ad38-aa2dbc367b29   Kazibwe Tonny   
1196  50130c50-99aa-48e6-ac49-7cb9b4c704e7  Mandeep Jareda   
1197  1e1abc05-53c6-4f21-9fbc-28dd5243b50f    chandra bhan   

                                                Comment  Thumbs_Up_Count  \
0     Twitter's obsession with turning itself into t...             7121   
1     Absolute rubbish. It still keeps refreshing af...             4833   
2     The most recent update has ruined this app. Th...             2409   
3     No closed caption (cc) button is present to di...             4033   
4     This app is frustrating as heck. When you're s...             1671   
...                                                 ...              ...   
1193                                              graet                1   
1194                                             I like                1   
1195                                          Great app                1   
1196                            5 star for Elon Musk ✌️                2   
1197                                           Nice app                0   

               Version Review_Created_At  
0     9.65.0-release.0  28-10-2022 01:48  
1     9.56.2-release.0  26-08-2022 08:26  
2     9.56.2-release.0  26-08-2022 18:31  
3     9.56.2-release.0  26-08-2022 16:05  
4     9.57.0-release.1  02-09-2022 07:37  
...                ...               ...  
1193  9.65.1-release.0  05-11-2022 11:56  
1194  9.65.1-release.0  05-11-2022 11:54  
1195                    05-11-2022 11:53  
1196  9.65.1-release.0  05-11-2022 11:53  
1197  9.61.0-release.0  05-11-2022 11:47  

[1198 rows x 6 columns]

In [ ]:
df2

reviewId  \
0     gp:AOqpTOEhZuqSqqWnaKRgv-9ABYdajFUB0WugPGh-SG-...   
1     gp:AOqpTOH0WP4IQKBZ2LrdNmFy_YmpPCVrV3diEU9KGm3...   
2     gp:AOqpTOEMCkJB8Iq1p-r9dPwnSYadA5BkPWTf32Z1azu...   
3     gp:AOqpTOGFrUWuKGycpje8kszj3uwHN6tU_fd4gLVFy9z...   
4     gp:AOqpTOHls7DW8wmDFzTkHwxuqFkdNQtKHmO6Pt9jhZE...   
...                                                 ...   
8615  gp:AOqpTOFp_TTAYjsFrPH-bq_Ag7bewBqv6AgGAPSJkf7...   
8616  gp:AOqpTOGH3IhuvYPoQqY5UN_tDcRDj3Xh1KS22uJL15M...   
8617  gp:AOqpTOENsTEUPKnBoNf8gpXid15yI7EjAwC7e_rpYF1...   
8618  gp:AOqpTOEmo0wzxI7QpduQpnKzj3Jo8fqKJD6jOCzUeG3...   
8619  gp:AOqpTOF6g4Dz-s82IINPpadB1whiiwUKPEuD9IJAoLM...   

                              userName  \
0                             Eric Tie   
1                           john alpha   
2                          Sudhakar .S   
3     SKGflorida@bellsouth.net DAVID S   
4                        Louann Stoker   
...                                ...   
8615                        Shine Or2z   
8616                       Chetan Iske   
8617                         Cloud CWG   
8618                      satish kumar   
8619                     Nithin Xavier   

                                              userImage  \
0     https://play-lh.googleusercontent.com/a-/AOh14...   
1     https://play-lh.googleusercontent.com/a-/AOh14...   
2     https://play-lh.googleusercontent.com/a-/AOh14...   
3     https://play-lh.googleusercontent.com/-75aK0WF...   
4     https://play-lh.googleusercontent.com/-pBcY_Z-...   
...                                                 ...   
8615  https://play-lh.googleusercontent.com/-3VnnzLh...   
8616  https://play-lh.googleusercontent.com/a-/AOh14...   
8617  https://play-lh.googleusercontent.com/a-/AOh14...   
8618  https://play-lh.googleusercontent.com/-lh-C7Cw...   
8619  https://play-lh.googleusercontent.com/a-/AOh14...   

                                                content  score  thumbsUpCount  \
0                         I cannot open the app anymore      1              0   
1     I have been begging for a refund from this app...      1              0   
2     Very costly for the premium version (approx In...      1              0   
3     Used to keep me organized, but all the 2020 UP...      1              0   
4                                   Dan Birthday Oct 28      1              0   
...                                                 ...    ...            ...   
8615  Uninstalled within 10 minutes of confusion. Wh...      1              0   
8616         Only notifications pop out no sound at all      1              0   
8617  If you want to use it properly, you have to pa...      1              0   
8618                                         Love it🥰🥰🥰      1              0   
8619  I paid for premium but I did not get any notif...      1              0   

     reviewCreatedVersion                   at  \
0                 5.4.0.6  2020-10-27 21:24:41   
1                          2020-10-27 14:03:28   
2                          2020-10-27 08:18:40   
3                          2020-10-26 13:28:07   
4                 5.6.0.7  2020-10-26 06:10:50   
...                   ...                  ...   
8615                       2020-06-30 02:54:42   
8616                 2.10  2020-06-29 18:55:33   
8617                       2020-06-29 09:15:59   
8618                 2.10  2020-06-26 13:48:16   
8619                 2.10  2020-06-22 23:22:50   

                                           replyContent            repliedAt  \
0                                                                              
1     Please note that from checking our records, yo...  2020-10-27 15:05:52   
2                                                                              
3     What do you find troublesome about the update?...  2020-10-26 14:58:29   
4                                                                              
...                                                 ...               

In [ ]:
#removing unwanted columns
df2 = df2.drop(columns=['userImage'])
df2 = df2.drop(columns=['replyContent'])
df2 = df2.drop(columns=['repliedAt'])
df2 = df2.drop(columns=['sortOrder'])

In [ ]:
df2 = df2.drop(columns=['appId'])

In [ ]:
df2

reviewId  \
0     gp:AOqpTOEhZuqSqqWnaKRgv-9ABYdajFUB0WugPGh-SG-...   
1     gp:AOqpTOH0WP4IQKBZ2LrdNmFy_YmpPCVrV3diEU9KGm3...   
2     gp:AOqpTOEMCkJB8Iq1p-r9dPwnSYadA5BkPWTf32Z1azu...   
3     gp:AOqpTOGFrUWuKGycpje8kszj3uwHN6tU_fd4gLVFy9z...   
4     gp:AOqpTOHls7DW8wmDFzTkHwxuqFkdNQtKHmO6Pt9jhZE...   
...                                                 ...   
8615  gp:AOqpTOFp_TTAYjsFrPH-bq_Ag7bewBqv6AgGAPSJkf7...   
8616  gp:AOqpTOGH3IhuvYPoQqY5UN_tDcRDj3Xh1KS22uJL15M...   
8617  gp:AOqpTOENsTEUPKnBoNf8gpXid15yI7EjAwC7e_rpYF1...   
8618  gp:AOqpTOEmo0wzxI7QpduQpnKzj3Jo8fqKJD6jOCzUeG3...   
8619  gp:AOqpTOF6g4Dz-s82IINPpadB1whiiwUKPEuD9IJAoLM...   

                              userName  \
0                             Eric Tie   
1                           john alpha   
2                          Sudhakar .S   
3     SKGflorida@bellsouth.net DAVID S   
4                        Louann Stoker   
...                                ...   
8615                        Shine Or2z   
8616                       Chetan Iske   
8617                         Cloud CWG   
8618                      satish kumar   
8619                     Nithin Xavier   

                                                content  score  thumbsUpCount  \
0                         I cannot open the app anymore      1              0   
1     I have been begging for a refund from this app...      1              0   
2     Very costly for the premium version (approx In...      1              0   
3     Used to keep me organized, but all the 2020 UP...      1              0   
4                                   Dan Birthday Oct 28      1              0   
...                                                 ...    ...            ...   
8615  Uninstalled within 10 minutes of confusion. Wh...      1              0   
8616         Only notifications pop out no sound at all      1              0   
8617  If you want to use it properly, you have to pa...      1              0   
8618                                         Love it🥰🥰🥰      1              0   
8619  I paid for premium but I did not get any notif...      1              0   

     reviewCreatedVersion                   at  
0                 5.4.0.6  2020-10-27 21:24:41  
1                          2020-10-27 14:03:28  
2                          2020-10-27 08:18:40  
3                          2020-10-26 13:28:07  
4                 5.6.0.7  2020-10-26 06:10:50  
...                   ...                  ...  
8615                       2020-06-30 02:54:42  
8616                 2.10  2020-06-29 18:55:33  
8617                       2020-06-29 09:15:59  
8618                 2.10  2020-06-26 13:48:16  
8619                 2.10  2020-06-22 23:22:50  

[8620 rows x 7 columns]

In [ ]:
#removing unwanted columns
df3 = df3.drop(columns=['userImage'])
df3 = df3.drop(columns=['replyContent'])
df3 = df3.drop(columns=['repliedAt'])

In [ ]:
df3

reviewId              userName  \
0     68ccaec8-1415-4301-a85e-3004679a3a83          Cassie Moore   
1     d84cbfd3-6aa3-485c-aaf9-c5dca27dc966         Kaleb Plummer   
2     96618aa1-31e5-4259-8649-89b75d962f00           Rylee Maher   
3     078c0bda-598b-474e-a04e-d7cb3e6f6301  Kittykatelyn Romilly   
4     8e68c5cd-b12a-4206-a8da-6bfdbff44ae3      Loveness Malenga   
...                                    ...                   ...   
1811  52230a10-51f5-4c94-9e21-5aa73084b9d7       Folad Folad jan   
1812  ea27e457-45d3-44ed-98dc-23805e6d82d7         Jeffrey Ewing   
1813  139cbce7-5980-4ffe-a6eb-0926087e91ff             haji saab   
1814  32b5abf6-6cb1-418b-87cf-39a8afbaa32e       Bridget Ibrahim   
1815  dea77f47-823c-45d5-b646-679eb579011a        AHSAN NIAZI571   

                                                content  score  thumbsUpCount  \
0                                              No words      5              0   
1                                 Great fun app so far!      5              0   
2     The app would get a higher rating but I litera...      1              0   
3     I WISH I COULD GIVE THIS A 100 PERCENT RATING ...      5              0   
4                                   Pictures and record      5              0   
...                                                 ...    ...            ...   
1811                                               خوبه      2              0   
1812  Hey maby get rid of all the nazis and racists ...      1              0   
1813                                             Foryou      5              0   
1814  i would have give it a five star but it very s...      4              0   
1815                       Please an freez my account 😔      5              0   

     reviewCreatedVersion                   at  
0                  27.1.3  2022-11-29 21:55:37  
1                          2022-11-29 21:55:04  
2                  27.1.3  2022-11-29 21:54:48  
3                          2022-11-29 21:54:35  
4                          2022-11-29 21:54:21  
...                   ...                  ...  
1811                       2022-11-29 05:58:59  
1812               27.1.3  2022-11-29 05:58:14  
1813               27.1.3  2022-11-29 05:58:01  
1814               27.0.3  2022-11-29 05:56:57  
1815               26.0.3  2022-11-29 05:56:35  

[1816 rows x 7 columns]

In [1]:
pip install --upgrade 'sqlalchemy<2.0' mysql-connector-python pandas ipython-sql

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Using cached ipython_sql-0.5.0-py3-none-any.whl (20 kB)


In [2]:
from sqlalchemy import create_engine
import pandas as pd
import mysql.connector
%load_ext sql

engine = create_engine('mysql://USERNAME:PASSWORD@HOST/DATABASE')

**The purpose of Query 1 is to retrieve the total number of reviews for each of the three apps - Twitter, Anydo and TikTok - from their respective tables in the database. The UNION ALL operator is used to combine the results of the three SELECT statements into a single table with columns for app name and total number of reviews.**

In [3]:
# Query 1 - Check the total number of reviews for each app.
query1 = """
SELECT 'Twitter' AS app_name, COUNT(*) AS total_reviews FROM `Twitter_Reviews`
UNION ALL
SELECT 'Anydo' AS app_name, COUNT(*) AS total_reviews FROM `anydo_playstore_reviews`
UNION ALL
SELECT 'TikTok' AS app_name, COUNT(*) AS total_reviews FROM `tiktok_google_play_reviews`;
"""
df1 = pd.read_sql_query(query1, engine)
display(df1)


app_name  total_reviews
0  Twitter           1198
1    Anydo           8620
2   TikTok           1816

The insight from this query is that TikTok has the highest number of reviews among the three apps, while Anydo has the lowest. The total number of reviews can give an indication of the popularity of the app and the level of engagement from its users.

**This query is finding the average thumbs up count and average rating for each app by app_name. It uses a UNION ALL to combine the results for each app into a single dataframe.**

In [4]:
# Query 2 - Find the average rating and the average thumbs up count for each app.
query2 = """
SELECT 'Twitter' AS app_name, AVG(`Thumbs_Up_Count`) AS avg_thumbs_up, NULL AS avg_rating FROM `Twitter_Reviews`
UNION ALL
SELECT 'Anydo' AS app_name, AVG(`thumbsUpCount`) AS avg_thumbs_up, AVG(`score`) AS avg_rating FROM `anydo_playstore_reviews`
UNION ALL
SELECT 'TikTok' AS app_name, AVG(`thumbsUpCount`) AS avg_thumbs_up, AVG(`score`) AS avg_rating FROM `tiktok_google_play_reviews`;
"""
df2 = pd.read_sql_query(query2, engine)
display(df2)

app_name  avg_thumbs_up  avg_rating
0  Twitter        166.298         NaN
1    Anydo          3.171      3.1005
2   TikTok          0.293      4.2439

The output shows the average thumbs up count and average rating for each app in separate columns. The Twitter app does not have a rating, so the column for avg_rating shows NULL. The Anydo and Tiktok apps have both an average thumbs up count and an average rating.

The insight we can gain from this query is that the Anydo and Tiktok apps have similar average thumbs up counts, but the Anydo app has a slightly higher average rating than the Tiktok app. The Twitter app has a much lower average thumbs up count than the other two apps, which suggests that it may not be as popular or well-liked as the other two.

**After running the query, we can see that Twitter has the highest percentage of reviews with thumbs-up count greater than or equal to 10 (30.05%), followed by Anydo (6.87%) and TikTok (0.17%). This insight is useful for understanding the level of engagement and satisfaction among users of each app, and can help the app developers identify areas for improvement.**

In [5]:
#Query 3 - Find the percentage of reviews with thumbs-up count greater than or equal to 10 
query3 = """
SELECT app_name, 
       SUM(CASE WHEN thumbs_up >= 10 THEN 1 ELSE 0 END) / COUNT(*) * 100 AS percentage_of_reviews_with_10_or_more_thumbs_up
FROM (
    SELECT 'Twitter' AS app_name, `Thumbs_Up_Count` AS thumbs_up FROM `Twitter_Reviews`
    UNION ALL
    SELECT 'Anydo' AS app_name, `thumbsUpCount` AS thumbs_up FROM `anydo_playstore_reviews`
    UNION ALL
    SELECT 'TikTok' AS app_name, `thumbsUpCount` AS thumbs_up FROM `tiktok_google_play_reviews`
) AS subquery
GROUP BY app_name;
"""
df3 = pd.read_sql_query(query3, engine)
display(df3)


app_name  percentage_of_reviews_with_10_or_more_thumbs_up
0  Twitter                                          30.0501
1    Anydo                                           6.8677
2   TikTok                                           0.1652

After running the query, we can see that Twitter has the highest percentage of reviews with thumbs-up count greater than or equal to 10 (30.05%), followed by Anydo (6.87%) and TikTok (0.17%). This insight is useful for understanding the level of engagement and satisfaction among users of each app, and can help the app developers identify areas for improvement.

**Query 4 creates a view called all_app_reviews_word_split that splits the review content into individual words. Then, it finds the most common words used in the reviews for each app, up to the top 5 per platform. The query first creates a CTE word_counts that calculates the count of each word used in the reviews for each app, and then assigns a row number to each group of words based on the count. The main query selects the words with the top 5 counts for each app, and orders them by the app name and row number.**

In [6]:
# Query 4 - Create a view that splits the review content into individual words
drop_view_sql_word_split = "DROP VIEW IF EXISTS all_app_reviews_word_split;"
with engine.connect() as con:
    con.execute(drop_view_sql_word_split)

create_view_sql_word_split = """
CREATE VIEW all_app_reviews_word_split AS
SELECT app_name,
       User_Name,
       Comment,
       Thumbs_Up_Count,
       Version,
       created_at,
       SUBSTRING_INDEX(SUBSTRING_INDEX(Comment, ' ', n), ' ', -1) AS word
FROM all_app_reviews
JOIN (SELECT 1 n UNION ALL SELECT 2 UNION ALL SELECT 3 UNION ALL SELECT 4 UNION ALL SELECT 5 UNION ALL SELECT 6 UNION ALL SELECT 7 UNION ALL SELECT 8 UNION ALL SELECT 9 UNION ALL SELECT 10) numbers
ON CHAR_LENGTH(Comment) - CHAR_LENGTH(REPLACE(Comment, ' ', '')) >= n - 1;
"""

with engine.connect() as con:
    con.execute(create_view_sql_word_split)

# Query 4 - Find the most common words used in the reviews for each app (top 5 per platform)
query4 = """
WITH word_counts AS (
  SELECT app_name, word, COUNT(*) AS word_count,
         ROW_NUMBER() OVER (PARTITION BY app_name ORDER BY COUNT(*) DESC) AS row_num
  FROM all_app_reviews_word_split
  WHERE word NOT IN ('', 'the', 'and', 'to', 'a', 'of', 'in', 'for', 'is', 'on', 'with', 'I', 'it', 'that', 'this', 'as', 'are', 'have', 'was', 'but', 'my', 'be', 'at', 'not', 'or', 'by', 'so', 'if', 'from', 'an', 'there', 'has', 'about', 'you', 'The', 'can', 'will', 'its', 'their', 'they', 'them', 'we', 'our', 'us', 'your', 'when', 'would', 'should', 'how', 'who', 'what', 'where', 'which', 'It''s','me','app','very','really','app.','twitter','tiktok')
  GROUP BY app_name, word
)
SELECT app_name, word, word_count
FROM word_counts
WHERE row_num <= 5
ORDER BY app_name, row_num;
"""

df4 = pd.read_sql_query(query4, engine)
display(df4)



<ipython-input-6-884abf8e8082>:4: RemovedIn20Warning: Deprecated API features detected! These feature(s) are not compatible with SQLAlchemy 2.0. To prevent incompatible upgrades prior to updating applications, ensure requirements files are pinned to "sqlalchemy<2.0". Set environment variable SQLALCHEMY_WARN_20=1 to show all deprecation warnings.  Set environment variable SQLALCHEMY_SILENCE_UBER_WARNING=1 to silence this message. (Background on SQLAlchemy 2.0 at: https://sqlalche.me/e/b8d9)
  con.execute(drop_view_sql_word_split)


app_name    word  word_count
0     Anydo    good         839
1     Anydo   great         606
2     Anydo    love         476
3     Anydo    like         465
4     Anydo     use         317
5    TikTok    Good         251
6    TikTok    nice         154
7    TikTok    love         145
8    TikTok   Great          95
9    TikTok    best          94
10  Twitter    good          86
11  Twitter  update          76
12  Twitter    like          59
13  Twitter    elon          54
14  Twitter     New          51

The fourth query creates a view that splits the review content into individual words, and then finds the most common words used in the reviews for each app (top 5 per platform). We can see that the word "good" is the most common in Anydo and TikTok reviews, while "good" and "update" are the most common in Twitter reviews. This information can be useful for each company to gain insights into their customer's preferences and needs.

**This query calculates the number of reviews for each app that fall into three categories based on the number of thumbs up counts: low (0-9), medium (10-19), and high (>=20). It uses a common table expression (CTE) to first categorize each review based on the thumbs up count, and then the main query aggregates the number of reviews for each app in each category.**

In [8]:
# Query 5 - Find the number of reviews per app that have a high (>= 20), medium (10-19), or low (0-9) thumbs-up count
query5 = """
WITH thumbs_up_categories AS (
    SELECT app_name,
           Thumbs_Up_Count AS thumbs_up,
           CASE
               WHEN Thumbs_Up_Count >= 20 THEN 'High'
               WHEN Thumbs_Up_Count BETWEEN 10 AND 19 THEN 'Medium'
               ELSE 'Low'
           END AS thumbs_up_category
    FROM all_app_reviews
)
SELECT app_name,
       thumbs_up_category,
       COUNT(*) AS num_reviews
FROM thumbs_up_categories
GROUP BY app_name, thumbs_up_category
ORDER BY app_name, thumbs_up_category DESC;
"""

df5 = pd.read_sql_query(query5, engine)
display(df5)


app_name thumbs_up_category  num_reviews
0    Anydo             Medium          256
1    Anydo                Low         8028
2    Anydo               High          336
3   TikTok                Low         1813
4   TikTok               High            3
5  Twitter             Medium           90
6  Twitter                Low          838
7  Twitter               High          270

We can see that Anydo has the highest number of low thumbs-up count reviews with 8028, followed by Twitter with 838 and TikTok with 1813. Twitter has the highest number of high thumbs-up count reviews with 270, followed by Anydo with 336 and TikTok with 3.

**The purpose of Query 6 is to identify the top 5 most frequently mentioned keywords in positive and negative comments for each app. The query creates a view named "comment_words" which splits the comment column of each app into individual words. The query then creates another view named "word_counts" which counts the number of positive and negative comments for each word in the comment_words view. Finally, the query selects the top 5 most frequently mentioned keywords in positive and negative comments for each app by using the RANK() function to assign a rank to each keyword based on the positive and negative counts.**

In [9]:
# Query 6 - Identify the top 5 most frequently mentioned keywords in positive and negative comments for each app
query6 = """
WITH comment_words AS (
  SELECT app_name,
         Thumbs_Up_Count,
         SUBSTRING_INDEX(SUBSTRING_INDEX(Comment, ' ', n.n), ' ', -1) AS word
  FROM all_app_reviews
  JOIN (
    SELECT 1 AS n
    UNION ALL SELECT 2 UNION ALL SELECT 3 UNION ALL SELECT 4 UNION ALL SELECT 5
    UNION ALL SELECT 6 UNION ALL SELECT 7 UNION ALL SELECT 8 UNION ALL SELECT 9
    UNION ALL SELECT 10
  ) n
  ON CHAR_LENGTH(Comment) - CHAR_LENGTH(REPLACE(Comment, ' ', '')) >= n.n - 1
),
word_counts AS (
  SELECT app_name,
         word,
         SUM(CASE WHEN Thumbs_Up_Count >= 3 THEN 1 ELSE 0 END) AS positive_count,
         SUM(CASE WHEN Thumbs_Up_Count < 3 THEN 1 ELSE 0 END) AS negative_count
  FROM comment_words
  WHERE word NOT IN ('', 'the', 'and', 'to', 'a', 'of', 'in', 'for', 'is', 'on', 'with', 'I', 'it', 'that', 'this', 'as', 'are', 'have', 'was', 'but', 'my', 'be', 'at', 'not', 'or', 'by', 'so', 'if', 'from', 'an', 'there', 'has', 'about', 'you', 'The', 'can', 'will', 'its', 'their', 'they', 'them', 'we', 'our', 'us', 'your', 'when', 'would', 'should', 'how', 'who', 'what', 'where', 'which', 'It''s', 'me', 'app', 'very', 'really', 'app.', 'twitter', 'tiktok')
  GROUP BY app_name, word
)
SELECT app_name,
       word,
       positive_count,
       negative_count
FROM (
  SELECT *,
         ROW_NUMBER() OVER (PARTITION BY app_name ORDER BY positive_count DESC) AS positive_rank,
         ROW_NUMBER() OVER (PARTITION BY app_name ORDER BY negative_count DESC) AS negative_rank
  FROM word_counts
) ranked_words
WHERE positive_rank <= 5 OR negative_rank <= 5
ORDER BY app_name, positive_count DESC, negative_count DESC;
"""
df6 = pd.read_sql_query(query6, engine)
display(df6)


app_name     word  positive_count  negative_count
0     Anydo     love           112.0           364.0
1     Anydo     like           104.0           361.0
2     Anydo    great           103.0           503.0
3     Anydo     good            84.0           755.0
4     Anydo     app,            75.0           198.0
5     Anydo     Nice            21.0           275.0
6    TikTok     Good             1.0           250.0
7    TikTok    don't             1.0            20.0
8    TikTok     Post             1.0             2.0
9    TikTok   point.             1.0             0.0
10   TikTok  certain             1.0             0.0
11   TikTok     nice             0.0           154.0
12   TikTok     love             0.0           145.0
13   TikTok    Great             0.0            95.0
14   TikTok     best             0.0            94.0
15  Twitter   update            59.0            17.0
16  Twitter      New            32.0            19.0
17  Twitter     like            28.0            31.0
18  Twitter     been            25.0            16.0
19  Twitter   recent            25.0             3.0
20  Twitter     good            21.0            65.0
21  Twitter    great            21.0            24.0
22  Twitter     elon            13.0            41.0
23  Twitter     musk             6.0            27.0

We can see that "love", "like", and "good" are commonly mentioned in positive reviews for Anydo. In negative reviews, "app" is commonly mentioned. For TikTok, we can see that "good" is mentioned in one positive review, and "don't" and "Post" are mentioned in one negative review each. For Twitter, "update" and "New" are commonly mentioned in positive reviews, while "app" is mentioned frequently in negative reviews.

**This query retrieves the distribution of review scores for each app in the dataset. The result includes the app name, review score, and the number of reviews for each score.**

In [10]:
# Query 7 - Distribution of review scores for each app
query7 = """
SELECT app_name,
       score AS review_score,
       COUNT(*) AS num_reviews
FROM all_app_reviews
GROUP BY app_name, review_score
ORDER BY app_name, review_score;
"""
df7 = pd.read_sql_query(query7, engine)
display(df7)


app_name  review_score  num_reviews
0     Anydo           1.0         1738
1     Anydo           2.0         1588
2     Anydo           3.0         1361
3     Anydo           4.0         1936
4     Anydo           5.0         1997
5    TikTok           1.0          249
6    TikTok           2.0           53
7    TikTok           3.0           61
8    TikTok           4.0           96
9    TikTok           5.0         1357
10  Twitter           NaN         1198

The insight from this query is the distribution of review scores for each app in the dataset. We can analyze the results to determine the overall sentiment of the reviews for each app.

**The query defined above creates a view named 'all_app_reviews' that consolidates the review data from multiple tables into a single view. It combines the relevant columns from the 'Twitter_Reviews', 'anydo_playstore_reviews', and 'tiktok_google_play_reviews' tables, and adds an additional column 'app_name' to identify the source app for each review.**

Query 8 utilizes the 'all_app_reviews' view to find the app versions with the most positive and negative feedback. It first calculates the sentiment of each review based on the score, categorizing them as 'positive', 'negative', or 'neutral'. Then, it counts the number of reviews for each app version and sentiment. Finally, it selects the app versions with the highest count for positive and negative sentiments for each app.

In [20]:
# Define the query to create the view 'all_app_reviews'
create_view_query = """
CREATE OR REPLACE VIEW all_app_reviews AS
SELECT
  Review_Id AS review_id,
  User_Name AS user_name,
  User_Image AS user_image,
  Comment AS content,
  NULL AS score,  -- Placeholder value for score column in Twitter_Reviews
  Thumbs_Up_Count AS thumbs_up_count,
  Version AS review_created_version,
  Review_Created_At AS at,
  Sort_Order AS sort_order,
  'Twitter' AS app_name
FROM Twitter_Reviews
UNION ALL
SELECT
  reviewId AS review_id,
  userName AS user_name,
  userImage AS user_image,
  content,
  score,
  thumbsUpCount AS thumbs_up_count,
  reviewCreatedVersion AS review_created_version,  -- Correct column name for app version
  at,
  sortOrder AS sort_order,
  'Anydo' AS app_name
FROM anydo_playstore_reviews
UNION ALL
SELECT
  reviewId AS review_id,
  userName AS user_name,
  userImage AS user_image,
  content,
  score,
  thumbsUpCount AS thumbs_up_count,
  reviewCreatedVersion AS review_created_version,  -- Correct column name for app version
  at,
  NULL AS sort_order,
  'TikTok' AS app_name
FROM tiktok_google_play_reviews;
"""

# Execute the query to create the view
engine.execute(create_view_query)

# Define Query 8 to find app versions with the most positive and negative feedback
query8 = """
WITH version_sentiment AS (
  SELECT app_name,
         review_created_version AS app_version,
         CASE 
           WHEN score >= 4 THEN 'positive'
           WHEN score <= 2 THEN 'negative'
           ELSE 'neutral'
         END AS sentiment,
         COUNT(*) AS sentiment_count
  FROM all_app_reviews
  WHERE review_created_version IS NOT NULL  -- Exclude rows with no app_version
  GROUP BY app_name, app_version, sentiment
)
SELECT app_name,
       app_version,
       sentiment,
       sentiment_count
FROM (
  SELECT *,
         RANK() OVER (PARTITION BY app_name, sentiment ORDER BY sentiment_count DESC) AS sentiment_rank
  FROM version_sentiment
) ranked
WHERE sentiment_rank = 1
ORDER BY app_name, sentiment;
"""

# Execute Query 8 and display the result
df8 = pd.read_sql_query(query8, engine)
display(df8)


app_name       app_version sentiment  sentiment_count
0    Anydo                    negative              934
1    Anydo                     neutral              245
2    Anydo                    positive              374
3   TikTok                    negative              111
4   TikTok            27.1.3   neutral               20
5   TikTok            27.1.3  positive              467
6  Twitter  9.65.1-release.0   neutral              390

The result of Query 8 displays the app name, app version, sentiment ('positive' or 'negative'), and the count of reviews with that sentiment for the respective app versions. This information helps identify the specific versions of the apps that have received the most positive and negative feedback.

This was the best varient i could get. the app version wouldn't translate due to poor datasets